src/libopensc/pkcs15-openpgp.c

```diff
@@ -529,26 +544,55 @@ sc_pkcs15emu_openpgp_init(sc_pkcs15_card_t *p15card)
 	if (r < 0)
 		goto failed;
 
-	/* If DO 7F21 holds data, we declare a cert object for pkcs15 */
-	if (file->size > 0) {
+	for(u8 i=0; i<PGP_MAX_NUM_CERTS; i++) {
 		struct sc_pkcs15_cert_info cert_info;
 		struct sc_pkcs15_object    cert_obj;
+		u8* buffer = malloc(MAX_OPENPGP_DO_SIZE);
+		size_t resp_len = 0;
+
+		if (buffer == NULL)
+			goto failed;
 
 		memset(&cert_info, 0, sizeof(cert_info));
 		memset(&cert_obj,  0, sizeof(cert_obj));
 
+		/* only try to SELECT DATA for OpenPGP >= v3 */
+		if (card->type >= SC_CARD_TYPE_OPENPGP_V3) {
+			r = sc_card_ctl(card, SC_CARDCTL_OPENPGP_SELECT_DATA, &i);
+			if (r < 0) {
+				LOG_TEST_RET(card->ctx, r, "Failed OpenPGP - select data");
+				goto failed;
+			}
+		}
+		sc_format_path(certs[i].path, &cert_info.path);
+
 		/* Certificate ID. We use the same ID as the authentication key */
-		cert_info.id.value[0] = 3;
-		cert_info.id.len = 1;
-		/* Authority, flag is zero */
-		/* The path following which PKCS15 will find the content of the object */
-		sc_format_path("3F007F21", &cert_info.path);
+		sc_pkcs15_format_id(certs[i].id, &cert_info.id);
+
+		resp_len = sc_get_data(card, 0x7F21, buffer, MAX_OPENPGP_DO_SIZE);
```

Comment:
thx, will update (change `resp_len` to `int` and catch `< 0`)

Reply:
done

Reply:
maybe the `free(buffer)` is not needed here, as other `goto failed` also omit this, any opinion on this?

Reply:
thanks for fixing the error checking.

`free`ing the buffer is actually needed to avoid a memory leak. And actually, if you go back to line 563 you also need to `free` `buffer` :wink: 

There also is the preemptive exit on line 597, where sc_pkcs15emu_add_x509_cert fails. Here, the error indicates that the function failed and with that, the responsibility to free the buffer is still in the hands of the caller. please add another `free` here as well. thank you!

Reply:
* `goto failed` line 563 -> `free(buffer)` added
* `goto failed` line 597 -> `free(buffer)` added

Reply:
line 564 returns on failure, I could remove the `goto failed` afterwards, right? 
because currently the `goto failed` will never be reached, or am I seeing this wrong?

Reply:
right. The goto on line 565 can not be reached now.

Reply:
removed it